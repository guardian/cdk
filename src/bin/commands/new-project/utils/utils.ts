import camelCase from "lodash.camelcase";
import upperFirst from "lodash.upperfirst";

export const pascalCase = (str: string): string => {
  return upperFirst(camelCase(str));
};

export interface Name {
  kebab: string;
  pascal: string;
}
