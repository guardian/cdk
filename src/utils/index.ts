export * as array from "./array";
export * as exec from "./exec";
export * as git from "./git";
export * as math from "./math";
export * as singleton from "./singleton";
