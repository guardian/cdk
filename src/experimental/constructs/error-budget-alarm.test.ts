import { Template } from "aws-cdk-lib/assertions";
import { MathExpression, Metric } from "aws-cdk-lib/aws-cloudwatch";
import { simpleGuStackForTesting } from "../../utils/test";
import { GuErrorBudgetAlarmExperimental } from "./error-budget-alarm";

describe("The ErrorBudgetAlarmExperimental construct", () => {
  it("should create the correct resources with basic config", () => {
    const stack = simpleGuStackForTesting();
    new GuErrorBudgetAlarmExperimental(stack, {
      sloName: "MapiFrontsAvailability",
      sloTarget: 0.999,
      badEvents: new Metric({ metricName: "HttpErrors", namespace: "TestLoadBalancerMetrics" }),
      validEvents: new Metric({ metricName: "HttpRequests", namespace: "TestLoadBalancerMetrics" }),
      snsTopicNameForAlerts: "test-sns-topic",
    });
    expect(Template.fromStack(stack).toJSON()).toMatchSnapshot();
  });

  it("should fail to create Alarm when slo target is above 99.95%", () => {
    const stack = simpleGuStackForTesting();
    expect(() => {
      new GuErrorBudgetAlarmExperimental(stack, {
        sloName: "MapiFrontsAvailability",
        sloTarget: 0.9996,
        badEvents: new Metric({ metricName: "HttpErrors", namespace: "TestLoadBalancerMetrics" }),
        validEvents: new Metric({ metricName: "HttpRequests", namespace: "TestLoadBalancerMetrics" }),
        snsTopicNameForAlerts: "test-sns-topic",
      });
    }).toThrowError("ErrorBudgetAlarm only works with SLO targets between 0.95 and 0.9995");
  });

  it("should fail to create Alarm when slo target is below 95%", () => {
    const stack = simpleGuStackForTesting();
    expect(() => {
      new GuErrorBudgetAlarmExperimental(stack, {
        sloName: "MapiFrontsAvailability",
        sloTarget: 0.9499,
        badEvents: new Metric({ metricName: "HttpErrors", namespace: "TestLoadBalancerMetrics" }),
        validEvents: new Metric({ metricName: "HttpRequests", namespace: "TestLoadBalancerMetrics" }),
        snsTopicNameForAlerts: "test-sns-topic",
      });
    }).toThrowError("ErrorBudgetAlarm only works with SLO targets between 0.95 and 0.9995");
  });

  it("should accept math expressions for more complicated definitions of bad/valid events", () => {
    const stack = simpleGuStackForTesting();
    new GuErrorBudgetAlarmExperimental(stack, {
      sloName: "MapiFrontsAvailability",
      sloTarget: 0.999,
      badEvents: new MathExpression({
        expression: "loadBalancerErrors + targetErrors",
        usingMetrics: {
          loadBalancerErrors: new Metric({ metricName: "LbHttpErrors", namespace: "TestLoadBalancerMetrics" }),
          targetErrors: new Metric({ metricName: "TargetHttpErrors", namespace: "TestTargetMetrics" }),
        },
      }),
      validEvents: new MathExpression({
        expression: "allRequests - invalidRequests",
        usingMetrics: {
          allRequests: new Metric({ metricName: "HttpRequests", namespace: "TestLoadBalancerMetrics" }),
          invalidRequests: new Metric({ metricName: "BadRequests", namespace: "TestLoadBalancerMetrics" }),
        },
      }),
      snsTopicNameForAlerts: "test-sns-topic",
    });
    expect(Template.fromStack(stack).toJSON()).toMatchSnapshot();
  });

  it("should support multiple GuErrorBudgetAlarmExperimentals in a single stack", () => {
    const stack = simpleGuStackForTesting();
    new GuErrorBudgetAlarmExperimental(stack, {
      sloName: "MapiFrontsAvailability",
      sloTarget: 0.999,
      badEvents: new Metric({ metricName: "HttpErrors", namespace: "TestLoadBalancerMetrics" }),
      validEvents: new Metric({ metricName: "HttpRequests", namespace: "TestLoadBalancerMetrics" }),
      snsTopicNameForAlerts: "test-sns-topic",
    });
    new GuErrorBudgetAlarmExperimental(stack, {
      sloName: "MapiFrontsLatency",
      sloTarget: 0.999,
      badEvents: new Metric({ metricName: "SlowResponses", namespace: "TestLoadBalancerMetrics" }),
      validEvents: new Metric({ metricName: "HttpRequests", namespace: "TestLoadBalancerMetrics" }),
      snsTopicNameForAlerts: "test-sns-topic",
    });
    // Each GuErrorBudgetAlarmExperimental creates 3 composite alarms (fast, medium, slow) so we expect 6 in total here
    Template.fromStack(stack).resourceCountIs("AWS::CloudWatch::CompositeAlarm", 6);
  });
});
