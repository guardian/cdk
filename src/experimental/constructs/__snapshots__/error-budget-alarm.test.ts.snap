// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ErrorBudgetAlarmExperimental construct should accept math expressions for more complicated definitions of bad/valid events 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuStack",
      "GuErrorBudgetAlarmExperimental",
    ],
    "gu:cdk:version": "TEST",
  },
  "Resources": {
    "ChildAlarmLongPeriodMapiFrontsAvailabilityFastBurnCompositeAlarmF1A44641": {
      "Properties": {
        "AlarmName": "ChildAlarmLongPeriodMapiFrontsAvailabilityFastBurnCompositeAlarm",
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "Metrics": [
          {
            "Expression": "badEvents/validEvents",
            "Id": "expr_1",
            "Label": "Observed failure rate",
          },
          {
            "Expression": "loadBalancerErrors + targetErrors",
            "Id": "badEvents",
            "ReturnData": false,
          },
          {
            "Id": "loadBalancerErrors",
            "MetricStat": {
              "Metric": {
                "MetricName": "LbHttpErrors",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 3600,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Id": "targetErrors",
            "MetricStat": {
              "Metric": {
                "MetricName": "TargetHttpErrors",
                "Namespace": "TestTargetMetrics",
              },
              "Period": 3600,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Expression": "allRequests - invalidRequests",
            "Id": "validEvents",
            "ReturnData": false,
          },
          {
            "Id": "allRequests",
            "MetricStat": {
              "Metric": {
                "MetricName": "HttpRequests",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 3600,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Id": "invalidRequests",
            "MetricStat": {
              "Metric": {
                "MetricName": "BadRequests",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 3600,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 0.014400000000000013,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ChildAlarmLongPeriodMapiFrontsAvailabilityMediumBurnCompositeAlarmB28EFC34": {
      "Properties": {
        "AlarmName": "ChildAlarmLongPeriodMapiFrontsAvailabilityMediumBurnCompositeAlarm",
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "Metrics": [
          {
            "Expression": "badEvents/validEvents",
            "Id": "expr_1",
            "Label": "Observed failure rate",
          },
          {
            "Expression": "loadBalancerErrors + targetErrors",
            "Id": "badEvents",
            "ReturnData": false,
          },
          {
            "Id": "loadBalancerErrors",
            "MetricStat": {
              "Metric": {
                "MetricName": "LbHttpErrors",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 21600,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Id": "targetErrors",
            "MetricStat": {
              "Metric": {
                "MetricName": "TargetHttpErrors",
                "Namespace": "TestTargetMetrics",
              },
              "Period": 21600,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Expression": "allRequests - invalidRequests",
            "Id": "validEvents",
            "ReturnData": false,
          },
          {
            "Id": "allRequests",
            "MetricStat": {
              "Metric": {
                "MetricName": "HttpRequests",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 21600,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Id": "invalidRequests",
            "MetricStat": {
              "Metric": {
                "MetricName": "BadRequests",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 21600,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 0.006000000000000005,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ChildAlarmLongPeriodMapiFrontsAvailabilitySlowBurnCompositeAlarm31625779": {
      "Properties": {
        "AlarmName": "ChildAlarmLongPeriodMapiFrontsAvailabilitySlowBurnCompositeAlarm",
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "Metrics": [
          {
            "Expression": "badEvents/validEvents",
            "Id": "expr_1",
            "Label": "Observed failure rate",
          },
          {
            "Expression": "loadBalancerErrors + targetErrors",
            "Id": "badEvents",
            "ReturnData": false,
          },
          {
            "Id": "loadBalancerErrors",
            "MetricStat": {
              "Metric": {
                "MetricName": "LbHttpErrors",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 86400,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Id": "targetErrors",
            "MetricStat": {
              "Metric": {
                "MetricName": "TargetHttpErrors",
                "Namespace": "TestTargetMetrics",
              },
              "Period": 86400,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Expression": "allRequests - invalidRequests",
            "Id": "validEvents",
            "ReturnData": false,
          },
          {
            "Id": "allRequests",
            "MetricStat": {
              "Metric": {
                "MetricName": "HttpRequests",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 86400,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Id": "invalidRequests",
            "MetricStat": {
              "Metric": {
                "MetricName": "BadRequests",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 86400,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 0.0030000000000000027,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ChildAlarmShortPeriodMapiFrontsAvailabilityFastBurnCompositeAlarmCF3AFF0A": {
      "Properties": {
        "AlarmName": "ChildAlarmShortPeriodMapiFrontsAvailabilityFastBurnCompositeAlarm",
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "Metrics": [
          {
            "Expression": "badEvents/validEvents",
            "Id": "expr_1",
            "Label": "Observed failure rate",
          },
          {
            "Expression": "loadBalancerErrors + targetErrors",
            "Id": "badEvents",
            "ReturnData": false,
          },
          {
            "Id": "loadBalancerErrors",
            "MetricStat": {
              "Metric": {
                "MetricName": "LbHttpErrors",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 300,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Id": "targetErrors",
            "MetricStat": {
              "Metric": {
                "MetricName": "TargetHttpErrors",
                "Namespace": "TestTargetMetrics",
              },
              "Period": 300,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Expression": "allRequests - invalidRequests",
            "Id": "validEvents",
            "ReturnData": false,
          },
          {
            "Id": "allRequests",
            "MetricStat": {
              "Metric": {
                "MetricName": "HttpRequests",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 300,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Id": "invalidRequests",
            "MetricStat": {
              "Metric": {
                "MetricName": "BadRequests",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 300,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 0.014400000000000013,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ChildAlarmShortPeriodMapiFrontsAvailabilityMediumBurnCompositeAlarmDBEC21EA": {
      "Properties": {
        "AlarmName": "ChildAlarmShortPeriodMapiFrontsAvailabilityMediumBurnCompositeAlarm",
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "Metrics": [
          {
            "Expression": "badEvents/validEvents",
            "Id": "expr_1",
            "Label": "Observed failure rate",
          },
          {
            "Expression": "loadBalancerErrors + targetErrors",
            "Id": "badEvents",
            "ReturnData": false,
          },
          {
            "Id": "loadBalancerErrors",
            "MetricStat": {
              "Metric": {
                "MetricName": "LbHttpErrors",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 1800,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Id": "targetErrors",
            "MetricStat": {
              "Metric": {
                "MetricName": "TargetHttpErrors",
                "Namespace": "TestTargetMetrics",
              },
              "Period": 1800,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Expression": "allRequests - invalidRequests",
            "Id": "validEvents",
            "ReturnData": false,
          },
          {
            "Id": "allRequests",
            "MetricStat": {
              "Metric": {
                "MetricName": "HttpRequests",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 1800,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Id": "invalidRequests",
            "MetricStat": {
              "Metric": {
                "MetricName": "BadRequests",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 1800,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 0.006000000000000005,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ChildAlarmShortPeriodMapiFrontsAvailabilitySlowBurnCompositeAlarmCA6C81C2": {
      "Properties": {
        "AlarmName": "ChildAlarmShortPeriodMapiFrontsAvailabilitySlowBurnCompositeAlarm",
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "Metrics": [
          {
            "Expression": "badEvents/validEvents",
            "Id": "expr_1",
            "Label": "Observed failure rate",
          },
          {
            "Expression": "loadBalancerErrors + targetErrors",
            "Id": "badEvents",
            "ReturnData": false,
          },
          {
            "Id": "loadBalancerErrors",
            "MetricStat": {
              "Metric": {
                "MetricName": "LbHttpErrors",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 7200,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Id": "targetErrors",
            "MetricStat": {
              "Metric": {
                "MetricName": "TargetHttpErrors",
                "Namespace": "TestTargetMetrics",
              },
              "Period": 7200,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Expression": "allRequests - invalidRequests",
            "Id": "validEvents",
            "ReturnData": false,
          },
          {
            "Id": "allRequests",
            "MetricStat": {
              "Metric": {
                "MetricName": "HttpRequests",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 7200,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Id": "invalidRequests",
            "MetricStat": {
              "Metric": {
                "MetricName": "BadRequests",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 7200,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 0.0030000000000000027,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "MapiFrontsAvailabilityFastBurnCompositeAlarmA8F736A8": {
      "Properties": {
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:sns:",
                {
                  "Ref": "AWS::Region",
                },
                ":",
                {
                  "Ref": "AWS::AccountId",
                },
                ":test-sns-topic",
              ],
            ],
          },
        ],
        "AlarmName": "TestMapiFrontsAvailabilityFastBurnCompositeAlarmF4979734",
        "AlarmRule": {
          "Fn::Join": [
            "",
            [
              "(ALARM("",
              {
                "Fn::GetAtt": [
                  "ChildAlarmLongPeriodMapiFrontsAvailabilityFastBurnCompositeAlarmF1A44641",
                  "Arn",
                ],
              },
              "") AND ALARM("",
              {
                "Fn::GetAtt": [
                  "ChildAlarmShortPeriodMapiFrontsAvailabilityFastBurnCompositeAlarmCF3AFF0A",
                  "Arn",
                ],
              },
              ""))",
            ],
          ],
        },
      },
      "Type": "AWS::CloudWatch::CompositeAlarm",
    },
    "MapiFrontsAvailabilityMediumBurnCompositeAlarmEC8DC379": {
      "Properties": {
        "ActionsSuppressor": {
          "Fn::GetAtt": [
            "MapiFrontsAvailabilityFastBurnCompositeAlarmA8F736A8",
            "Arn",
          ],
        },
        "ActionsSuppressorExtensionPeriod": 1800,
        "ActionsSuppressorWaitPeriod": 120,
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:sns:",
                {
                  "Ref": "AWS::Region",
                },
                ":",
                {
                  "Ref": "AWS::AccountId",
                },
                ":test-sns-topic",
              ],
            ],
          },
        ],
        "AlarmName": "TestMapiFrontsAvailabilityMediumBurnCompositeAlarm7022D40E",
        "AlarmRule": {
          "Fn::Join": [
            "",
            [
              "(ALARM("",
              {
                "Fn::GetAtt": [
                  "ChildAlarmLongPeriodMapiFrontsAvailabilityMediumBurnCompositeAlarmB28EFC34",
                  "Arn",
                ],
              },
              "") AND ALARM("",
              {
                "Fn::GetAtt": [
                  "ChildAlarmShortPeriodMapiFrontsAvailabilityMediumBurnCompositeAlarmDBEC21EA",
                  "Arn",
                ],
              },
              ""))",
            ],
          ],
        },
      },
      "Type": "AWS::CloudWatch::CompositeAlarm",
    },
    "MapiFrontsAvailabilitySlowBurnCompositeAlarm1198F7B4": {
      "Properties": {
        "ActionsSuppressor": {
          "Fn::GetAtt": [
            "MapiFrontsAvailabilityMediumBurnCompositeAlarmEC8DC379",
            "Arn",
          ],
        },
        "ActionsSuppressorExtensionPeriod": 7200,
        "ActionsSuppressorWaitPeriod": 120,
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:sns:",
                {
                  "Ref": "AWS::Region",
                },
                ":",
                {
                  "Ref": "AWS::AccountId",
                },
                ":test-sns-topic",
              ],
            ],
          },
        ],
        "AlarmName": "TestMapiFrontsAvailabilitySlowBurnCompositeAlarm5EDECA5E",
        "AlarmRule": {
          "Fn::Join": [
            "",
            [
              "(ALARM("",
              {
                "Fn::GetAtt": [
                  "ChildAlarmLongPeriodMapiFrontsAvailabilitySlowBurnCompositeAlarm31625779",
                  "Arn",
                ],
              },
              "") AND ALARM("",
              {
                "Fn::GetAtt": [
                  "ChildAlarmShortPeriodMapiFrontsAvailabilitySlowBurnCompositeAlarmCA6C81C2",
                  "Arn",
                ],
              },
              ""))",
            ],
          ],
        },
      },
      "Type": "AWS::CloudWatch::CompositeAlarm",
    },
  },
}
`;

exports[`The ErrorBudgetAlarmExperimental construct should create the correct resources with basic config 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuStack",
      "GuErrorBudgetAlarmExperimental",
    ],
    "gu:cdk:version": "TEST",
  },
  "Resources": {
    "ChildAlarmLongPeriodMapiFrontsAvailabilityFastBurnCompositeAlarmF1A44641": {
      "Properties": {
        "AlarmName": "ChildAlarmLongPeriodMapiFrontsAvailabilityFastBurnCompositeAlarm",
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "Metrics": [
          {
            "Expression": "badEvents/validEvents",
            "Id": "expr_1",
            "Label": "Observed failure rate",
          },
          {
            "Id": "badEvents",
            "MetricStat": {
              "Metric": {
                "MetricName": "HttpErrors",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 3600,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Id": "validEvents",
            "MetricStat": {
              "Metric": {
                "MetricName": "HttpRequests",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 3600,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 0.014400000000000013,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ChildAlarmLongPeriodMapiFrontsAvailabilityMediumBurnCompositeAlarmB28EFC34": {
      "Properties": {
        "AlarmName": "ChildAlarmLongPeriodMapiFrontsAvailabilityMediumBurnCompositeAlarm",
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "Metrics": [
          {
            "Expression": "badEvents/validEvents",
            "Id": "expr_1",
            "Label": "Observed failure rate",
          },
          {
            "Id": "badEvents",
            "MetricStat": {
              "Metric": {
                "MetricName": "HttpErrors",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 21600,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Id": "validEvents",
            "MetricStat": {
              "Metric": {
                "MetricName": "HttpRequests",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 21600,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 0.006000000000000005,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ChildAlarmLongPeriodMapiFrontsAvailabilitySlowBurnCompositeAlarm31625779": {
      "Properties": {
        "AlarmName": "ChildAlarmLongPeriodMapiFrontsAvailabilitySlowBurnCompositeAlarm",
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "Metrics": [
          {
            "Expression": "badEvents/validEvents",
            "Id": "expr_1",
            "Label": "Observed failure rate",
          },
          {
            "Id": "badEvents",
            "MetricStat": {
              "Metric": {
                "MetricName": "HttpErrors",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 86400,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Id": "validEvents",
            "MetricStat": {
              "Metric": {
                "MetricName": "HttpRequests",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 86400,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 0.0030000000000000027,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ChildAlarmShortPeriodMapiFrontsAvailabilityFastBurnCompositeAlarmCF3AFF0A": {
      "Properties": {
        "AlarmName": "ChildAlarmShortPeriodMapiFrontsAvailabilityFastBurnCompositeAlarm",
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "Metrics": [
          {
            "Expression": "badEvents/validEvents",
            "Id": "expr_1",
            "Label": "Observed failure rate",
          },
          {
            "Id": "badEvents",
            "MetricStat": {
              "Metric": {
                "MetricName": "HttpErrors",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 300,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Id": "validEvents",
            "MetricStat": {
              "Metric": {
                "MetricName": "HttpRequests",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 300,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 0.014400000000000013,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ChildAlarmShortPeriodMapiFrontsAvailabilityMediumBurnCompositeAlarmDBEC21EA": {
      "Properties": {
        "AlarmName": "ChildAlarmShortPeriodMapiFrontsAvailabilityMediumBurnCompositeAlarm",
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "Metrics": [
          {
            "Expression": "badEvents/validEvents",
            "Id": "expr_1",
            "Label": "Observed failure rate",
          },
          {
            "Id": "badEvents",
            "MetricStat": {
              "Metric": {
                "MetricName": "HttpErrors",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 1800,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Id": "validEvents",
            "MetricStat": {
              "Metric": {
                "MetricName": "HttpRequests",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 1800,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 0.006000000000000005,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ChildAlarmShortPeriodMapiFrontsAvailabilitySlowBurnCompositeAlarmCA6C81C2": {
      "Properties": {
        "AlarmName": "ChildAlarmShortPeriodMapiFrontsAvailabilitySlowBurnCompositeAlarm",
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "Metrics": [
          {
            "Expression": "badEvents/validEvents",
            "Id": "expr_1",
            "Label": "Observed failure rate",
          },
          {
            "Id": "badEvents",
            "MetricStat": {
              "Metric": {
                "MetricName": "HttpErrors",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 7200,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Id": "validEvents",
            "MetricStat": {
              "Metric": {
                "MetricName": "HttpRequests",
                "Namespace": "TestLoadBalancerMetrics",
              },
              "Period": 7200,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 0.0030000000000000027,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "MapiFrontsAvailabilityFastBurnCompositeAlarmA8F736A8": {
      "Properties": {
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:sns:",
                {
                  "Ref": "AWS::Region",
                },
                ":",
                {
                  "Ref": "AWS::AccountId",
                },
                ":test-sns-topic",
              ],
            ],
          },
        ],
        "AlarmName": "TestMapiFrontsAvailabilityFastBurnCompositeAlarmF4979734",
        "AlarmRule": {
          "Fn::Join": [
            "",
            [
              "(ALARM("",
              {
                "Fn::GetAtt": [
                  "ChildAlarmLongPeriodMapiFrontsAvailabilityFastBurnCompositeAlarmF1A44641",
                  "Arn",
                ],
              },
              "") AND ALARM("",
              {
                "Fn::GetAtt": [
                  "ChildAlarmShortPeriodMapiFrontsAvailabilityFastBurnCompositeAlarmCF3AFF0A",
                  "Arn",
                ],
              },
              ""))",
            ],
          ],
        },
      },
      "Type": "AWS::CloudWatch::CompositeAlarm",
    },
    "MapiFrontsAvailabilityMediumBurnCompositeAlarmEC8DC379": {
      "Properties": {
        "ActionsSuppressor": {
          "Fn::GetAtt": [
            "MapiFrontsAvailabilityFastBurnCompositeAlarmA8F736A8",
            "Arn",
          ],
        },
        "ActionsSuppressorExtensionPeriod": 1800,
        "ActionsSuppressorWaitPeriod": 120,
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:sns:",
                {
                  "Ref": "AWS::Region",
                },
                ":",
                {
                  "Ref": "AWS::AccountId",
                },
                ":test-sns-topic",
              ],
            ],
          },
        ],
        "AlarmName": "TestMapiFrontsAvailabilityMediumBurnCompositeAlarm7022D40E",
        "AlarmRule": {
          "Fn::Join": [
            "",
            [
              "(ALARM("",
              {
                "Fn::GetAtt": [
                  "ChildAlarmLongPeriodMapiFrontsAvailabilityMediumBurnCompositeAlarmB28EFC34",
                  "Arn",
                ],
              },
              "") AND ALARM("",
              {
                "Fn::GetAtt": [
                  "ChildAlarmShortPeriodMapiFrontsAvailabilityMediumBurnCompositeAlarmDBEC21EA",
                  "Arn",
                ],
              },
              ""))",
            ],
          ],
        },
      },
      "Type": "AWS::CloudWatch::CompositeAlarm",
    },
    "MapiFrontsAvailabilitySlowBurnCompositeAlarm1198F7B4": {
      "Properties": {
        "ActionsSuppressor": {
          "Fn::GetAtt": [
            "MapiFrontsAvailabilityMediumBurnCompositeAlarmEC8DC379",
            "Arn",
          ],
        },
        "ActionsSuppressorExtensionPeriod": 7200,
        "ActionsSuppressorWaitPeriod": 120,
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:sns:",
                {
                  "Ref": "AWS::Region",
                },
                ":",
                {
                  "Ref": "AWS::AccountId",
                },
                ":test-sns-topic",
              ],
            ],
          },
        ],
        "AlarmName": "TestMapiFrontsAvailabilitySlowBurnCompositeAlarm5EDECA5E",
        "AlarmRule": {
          "Fn::Join": [
            "",
            [
              "(ALARM("",
              {
                "Fn::GetAtt": [
                  "ChildAlarmLongPeriodMapiFrontsAvailabilitySlowBurnCompositeAlarm31625779",
                  "Arn",
                ],
              },
              "") AND ALARM("",
              {
                "Fn::GetAtt": [
                  "ChildAlarmShortPeriodMapiFrontsAvailabilitySlowBurnCompositeAlarmCA6C81C2",
                  "Arn",
                ],
              },
              ""))",
            ],
          ],
        },
      },
      "Type": "AWS::CloudWatch::CompositeAlarm",
    },
  },
}
`;
